### ISSUE TITLE
You can instantiate a dyn Trait that doesn't implement Trait

### ISSUE THREAD

**Issue by QuineDot** [rust-lang/rust#88904]
[I tried this code:](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=96212284062f0a38070c9ac9b2629343)

```rust
pub trait Marker<'a> {}

pub trait Trait where for<'a> &'a Self: Marker<'a> {
    fn usable(&self) {}
}

impl<'a> Marker<'a> for &'a () {}
impl Trait for () {}

pub fn f() {
    // `(): Trait`
    let u = ();
    u.usable();

    // `dyn Trait` exists and `()` can coerce to it...
    let _t: &dyn Trait = &();
    // But `dyn Trait: !Trait`
    // _t.usable();
    // error[E0277]: the trait bound `for<'a> &'a T: Marker<'a>` is not satisfied
}
```

I expected to see this happen: The unsized coercion is denied (but see discussion below).

Instead, this happened: The unsized coercion happened and we end up with a `dyn Trait` which does not implement `Trait`.

### Meta
This works on [any stable version of Rust](https://rust.godbolt.org/z/vsW5nv9es).

### Notes

The failure to meet the trait bound is caught when you actually try to _use_ the `dyn Trait`, at least trivially.  I haven't found a way to weaponize this, but I haven't put any effort into doing so either.

`dyn Trait` implements `Trait` if you `impl<'a> Marker<'a> for &'a (dyn Trait + 'static) {}` in the example.  However, the non-super-trait bound need not be a marker trait (it can have non-defaulted methods).

Related: [Implied Bounds](https://rust-lang.github.io/rfcs/2089-implied-bounds.html), #20671 

### Discussion

I'm filing this as a bug because the assertion that `dyn Trait: Trait` always holds has been expressed as a belief by Rust team members, e.g.
* [This comment by Niko](https://github.com/rust-lang/rust/issues/42235#issuecomment-305835485)
* [This section of RFC 2027](https://github.com/withoutboats/rfcs/blob/object-safe-for-dispatch/text/2027-object_safe_for_dispatch.md#detailed-design)

There are some possibilities in line with this assertion:
* Unsized coercion checks the bounds and is denied in this case
* Bounds that aren't super-traits make a trait object unsafe generally
* Bounds that aren't super-traits don't apply to `dyn Trait`

However, another possibility is to generalize RFC 2027 and just accept the current behavior:
* `dyn Trait` is always a valid type, as per RFC 2027
* Unsized coercion is still dependent on all methods being either object safe or not applicable (`where Self: Sized`)
  * Though this could be relaxed too (if you're not object safe, `dyn Trait` just doesn't implement `Trait`)
* If unsized coercion is possible, you can instantiate a `dyn Trait`, even if it doesn't implement `Trait`
* Naturally, you can't use the `Trait` methods if it doesn't implement `Trait`

I'm sure there are a lot of subtleties and backwards-compatibility concerns in any of the possible approaches.

**Comment by Jules-Bertholet** [comment]
Another method for doing this:

```rust
trait Super {
    type Assoc: ?Sized;
}

trait Sub: Super<Assoc = Self> {}

fn foo<T: Sub>(a: &T) {
    let b: &dyn Sub<Assoc = T> = a; // This line compiles
    bar(b) // This line does not
}

fn bar<T: Sub + ?Sized>(a: &T) {}
```
